enhancements/cluster-logging/hypershift-audit-logs.md

```diff
@@ -0,0 +1,219 @@
+
+The HLO only generates configuration.
+The CLO in the management cluster does the log collection and forwarding.
+
+### API Extensions
+
+#### HypershiftLogForwarder
+
+The `HyperShiftLogForwarder` resource is created on the _guest_ cluster, but is reconciled by
+the _hypershift logging operator_ on the _management_ cluster.
+
+This allows the user to specify audit forwarding configuration, but the management
+cluster controls if/how it is implemented in the HCP.
+
+`HyperShiftLogForwarder` is a subset of the `ClusterLogForwarder` API:
+- Only the `audit` input is allowed, and only API server audit logs will be forwarded.
+- There may be restrictions on the set of outputs supported from the management cluster.
+  - Possibly only a predefined set of named outputs will be allowed.
+- There may be restrictions on pipeline features that are not relevant to API audit logs.
+
+### Implementation Details
+
+#### ClusterLogForwarder multiple instances
+
+The ClusterLogForwarder is currently a singleton.
+To provide isolation between HCPs we need to deploy a forwarder per HCP namespace.
+This requires modifications to the CLO 
+
+_FIXME JIRA ref._
+
+#### ClusterLogForwarder HTTP listener
+
+The ClusterLogForwarder needs to be extended to act a a HTTP listener,
+so it can receive logs as a web hook from the API server in the HCP.
+
+_FIXME JIRA reference_
+
+_FIXME Is there a way to use the file system and avoid this requirement?_
+
+#### Hypershift Logging Operator
+
+This operator must
+- watch resources in a HCP API server, update their status 
+  - `HypershiftLogForwarder` 
+  - `Policy.audit.k8s.io`
+- create resources in the management cluster API server
+  - `ClusterLogForwarder`
+  - `Policy.audit.k8s.io`
+
+The policy objects can be copied verbatim. 
+
+The `HypershiftLogForwarder` fields are compatible with `ClusterLogForwarder` fields but some changes
+and validation may be needed:
+- validate that the forwarding request is valid and allowed.
+- possibly replace named outputs with a suitable output configuration;
+  if we want to restrict the user to a fixed set of alternatives
+- possibly add credentials derived from the guest cluster or the management cluster.
+  
+### Risks and Mitigations
+
+CPU/Network cost of extra log forwarding
+- We have to do this anyway, but user might choose a more verbose policy.
+- We could restrict the available logs at the API server to control this.
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions
+
+- Is there any way to write the audit logs to a file system that the log forwarder can read?
+  This would be simpler & more reliable than a web hook.
+- Can an operator on the management cluster reconcile and update status on resources in the guest cluster directly?
+  The API server is running in the HCP on the management cluster, but need to check for possible snags.
+- Should we allow the user to specify any destination for logs from the management cluster? If not:
```

Comment:
In what ways would we need to restrict destinations?  The limitations will need to be on what network paths are available from the management cluster.  The MC can talk to the destination directly.  Configuration of any network paths by the service consumer to facilitate this must be out of scope of this enhancement.

Reply:
It is very possible there are no restrictions needed. I had a vague idea that there might be connection issues from "different" clusters, but you are correct that if the destination is reachable from the MC then there's no reason we can't forward to it.